pkg/apis/autoscaling/validation/validation.go

```diff
@@ -282,7 +285,14 @@ func validateMetricSpec(spec autoscaling.MetricSpec, fldPath *field.Path) field.
 		}
 	}
 
-	expectedField := strings.ToLower(string(spec.Type))
+	if spec.ContainerResource != nil {
+		typesPresent.Insert("containerResource")
+		if typesPresent.Len() == 1 {
+			allErrs = append(allErrs, validateContainerResourceSource(spec.ContainerResource, fldPath.Child("containerResource"))...)
+		}
+	}
+
+	expectedField := getMetricExpectedField(spec.Type)
 
 	if !typesPresent.Has(expectedField) {
 		allErrs = append(allErrs, field.Required(fldPath.Child(expectedField), "must populate information for the given metric source"))
```

Comment:
can handle the invalid type and missing required field cases together here and make this more straightforward, in addition to providing a nicer message when the container resource feature is enabled:

```go
switch spec.Type {

case autoscaling.ObjectMetricSourceType:
  if spec.Object == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("object"), "must populate information for the given metric source"))
  }

case autoscaling.PodsMetricSourceType:
  if spec.Pods == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("pods"), "must populate information for the given metric source"))
  }

case autoscaling.ResourceMetricSourceType:
  if spec.Resource == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("resource"), "must populate information for the given metric source"))
  }

case autoscaling.ExternalMetricSourceType:
  if spec.External == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("external"), "must populate information for the given metric source"))
  }

case autoscaling.ContainerResourceMetricSourceType:
  if spec.ContainerResource == nil {
    if utilfeature.DefaultFeatureGate.Enabled(features.HPAContainerMetrics) {
      allErrs = append(allErrs, field.Required(fldPath.Child("containerResource"), "must populate information for the given metric source"))
    } else {
      allErrs = append(allErrs, field.Required(fldPath.Child("containerResource"), "must populate information for the given metric source (only allowed when HPAContainerMetrics feature is enabled)"))
    }
  }

default:
  allErrs = append(allErrs, field.NotSupported(fldPath.Child("type"), *hook.MatchPolicy, supportedMatchPolicies.List())
}
```